### DISCUSSION TITLE
How to train my own model

### DISCUSSION THREAD

**Discussion by Alex-joomla** [jomjol/AI-on-the-edge-device#2337] [Q&A]
As I'm already invested uncountable hours to get this thing running smoothly the last chance I see is to train my own model. I have the images extracted and renamed with the py tool you provided. But now I'm stuck. I tried the autotrain thng but got errors from the github action. Tried to download jupyter notebook but also errors. Is there somewhere a simple howto to do this :-/

Thank you guys

**Comment by caco3** [comment]
I guess you already studied the documentation (https://jomjol.github.io/AI-on-the-edge-device-docs/Learn-models-with-your-own-images/)

Did you also check if the mentioned repos in there contain additional information?

And last, you should provide the error messages, else its impossible to help.

**Reply by Alex-joomla** [reply]
First I tried the local installation of jupyter notebook on my max terminal and run the commands as listed in the link you shared. But the commands are too little described for me :-/ as I'm not a professional coder :-)

If I run 'jupyter notebook Image_Preparation.ipynb'

I get the error:
```
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In[1], line 1
----> 1 files = glob.glob(Input_dir + '/*.jpg')
      2 files = files + glob.glob(Input_dir + '/*.png')
      3 files = files + glob.glob(Input_dir + '/*.bmp')

NameError: name 'glob' is not defined
```

Then i searched further an tried `neural-network-autotrain-digital-counter`

Forked it (first time I "forked" something besides of spaghetti or hay in my life), uploaded my pictures. And thought the github action will du the rest. But got errors:
`[A job to say hello](https://github.com/Alex-joomla/neural-network-autotrain-digital-counter/actions/runs/4744428166/jobs/8425318377)
Node.js 12 actions are deprecated. Please update the following actions to use Node.js 16: actions/checkout@v2. For more information see: https://github.blog/changelog/2022-09-22-github-actions-all-actions-will-begin-running-on-node16-instead-of-node12/.`

**Reply by caco3** [reply]
@haverland @jomjol Any Idea why it does not work anymore?

I never used it myself.

**Reply by ghost** [reply]
Do you really need a dig-class11 model? The https://github.com/jomjol/neural-network-digital-counter-readout is for dig-class11 models. Newer models using https://github.com/haverland/Tenth-of-step-of-a-meter-digit. 
A preparation of images is not needed here any more. Best is to create a new folder in images/collected and save the your images here. 
Run the notebook [dig-class100-s2.ipynb](https://github.com/haverland/Tenth-of-step-of-a-meter-digit/blob/master/dig-class100-s2.ipynb) or [dig_class100-s1.ipynb](https://github.com/haverland/Tenth-of-step-of-a-meter-digit/blob/master/dig_class100-s1.ipynb) (bigger but more accuracy).

I will check the https://github.com/jomjol/neural-network-digital-counter-readout at weekend.

**Comment by hilsonp** [comment]
Just in case it helps anyone.
I wrote a cookbook for dummies (like me) showing how to train a model on Windows.
My 2 meters are now read perfectly by the 2 tiny ESB32-CAM.

Awesome work @jomjol and @haverland !
Thank you !!!

**Reply by Alex-joomla** [reply]
Would you be so kind and share it?

**Reply by hilsonp** [reply]
Sorry. I forgot to paste the link. 

Here it is: https://github.com/haverland/Tenth-of-step-of-a-meter-digit/issues/12

**Reply by ghost** [reply]
Thanks. I use your explanation for a short wiki page https://github.com/haverland/Tenth-of-step-of-a-meter-digit/wiki/Cookbook-running-the-jupyter-notebook-with-my-own-data

**Reply by ghost** [reply]
Added it to FAQ.

**Reply by gekkehenkie11** [reply]
Hey @haverland I just started with this, couldnt run the commands at first until I found out I needed x64 bits version of python LOL download it here: https://www.python.org/downloads/windows/
Then instead of just 'pip' my command line would need to start with 'python -m pip'. Probably obvious to most but python newbies like me might struggle with this already LOL. Thanks!

**Reply by ghost** [reply]
It's sometimes different depending on the default installation. If you have multiple installations, it's better to use python -m pip. But on Linux, Mac pip should be do the same. On windows I'm not familiar with.

**Comment by Alex-joomla** [comment]
~~@hilsonp the link isn't your cookbook isn't it?~~

found it thank you 

@haverland in my Desperation I tried to learn a new models with the slightly corrected ROI Config. But all I get out is a bucket of errors:

I'm starting the notebook in my mac-terminal with:
`jupyter notebook dig_class100-s1.ipynb`

Then I click a few times on the play button but I get:

<img width="1145" alt="Bildschirmfoto 2023-06-06 um 09 16 50" src="https://github.com/jomjol/AI-on-the-edge-device/assets/32268233/762f6243-e474-44a4-afc6-961884ec7a87">

What I did already was `pip install b2`but no success :-/

**Reply by ghost** [reply]
b2n is a directory in the root of Tenth-of-step-of-a-meter-digit and should normally be found it it started in project-root.

**Reply by ghost** [reply]
I've added __init__.py for better detection.

**Reply by ghost** [reply]
Oh Mac...

python --version 

is >= 3.9?

**Comment by gekkehenkie11** [comment]
Ok tried it, like said above it at first failed, found out that i needed to install the 64 bit version of python and call pip with 'python -m pip' instead of just 'pip'. Then it worked. Now I'll have to wait 3 days to collect the data LOL.

**Comment by gekkehenkie11** [comment]
@haverland  @hilsonp 
Ok one thing I dont have clear is: how many samples should I label ? Since I have my cam configured to take pictures every minute and I have 7 digits, this results in 420 samples per hour, 10,080 per day, 30,240 for the whole 3 days LOL.

So how to go about this? I guess the idea is to order them and then just find 100 samples (0.0 to 9.9) ? So you dont' really need to go over all of them?

Hmm then again, my last 3 digits are red, the others are black, so I guess that just for red I need 100 samples and then for black too ... So I guess 200 samples in my case makes sense?

Thanks!

**Reply by ghost** [reply]
May depend on the difference to the images in the dataset. But once for each (100 samples) should be good enough. To much samples of one step is not good. Means if you have hundreds of 0.0 please reduce the size. But this should be reduced by the [collectmeterdigits](https://github.com/haverland/collectmeterdigits) tool already.

Mostly you will not get once for each. If you have 100 - 200 collected test it with https://github.com/haverland/Tenth-of-step-of-a-meter-digit.

**Reply by gekkehenkie11** [reply]
cool! Thanks for your answer! Just started to collect and will experiment with this in 3 days. Thanks again!

**Comment by gekkehenkie11** [comment]
Ok tomorrow morning I'll have my 3 days of data, can't wait to start! @haverland one thing I'm not sure about: if I'm building a model just for my watermeter, wouldn't it make sense to delete ALL other data from your directory? so empty the whole 'collected' directory? Or even the whole 'images' directory? And then just paste my 100-200 labeled photos into it? Cause now for example it sometimes recognizes a 7 as a 1, so it must be getting that from one of the other photos. If I just add a directory with my own data (so keep the old data too), the model still might tend to choose based on the exisiting (old) data? Or doesnt it work like that?

Thanks!

**Reply by ghost** [reply]
Use all the data! Only your data would not enough and will tents to overfitting the model.

**Comment by gekkehenkie11** [comment]
Awesome, thanks! So to help out the next guy doing this, here's what I found and did: 

When downloading all the images from the device it halted several times, because of a bad connection (I dont have a 100% stable wifi at the meter). So I just restarted it several times with the keepdownloads option:

python -m collectmeterdigits --collect=192.168.0.100 --days=3 --keepdownloads

And it seemed to resume correctly. At the end the image processor window popped up. I numbered all images with it. It first deleted all similar images and then ended up with about 1500 images which I had to label. Probably took me about 2 hours, it's relatively easy with that tool.

Then next step there's a small typo, it's obvious to even me but for the record: 
git clone haverland/Tenth-of-step-of-a-meter-digit should of course be 

git clone https://github.com/haverland/Tenth-of-step-of-a-meter-digit

Then missing from the haverland wiki, but as noted by @hilsonp

pip install notebook
pip install opencv-python
pip install tensorflow-model-optimization

*EDIT* hmm the above is also in the requirements.txt from @haverland so, instead of running all those commands, you should just download and paste that requirements.txt and then process it, like haverland suggested
*EDIT2* hmm that "tensorflow-model-optimization" exists with an error "ERROR: Failed building wheel for dm-tree" , going to investigate
*EDIT3* Ok that error is because of the python version, you need v3.11. So on windows:

winget install python.python.3.11

and then call specifically 3.11 like this:

py -3.11 -m pip

so for example:

py -3.11 -m pip install  tensorflow-model-optimization

Or even better, just use the requirements. txt:

py -3.11 -m pip install -r requirements.txt

This installed everything correctly!

and then lastly a small typo in haverlands wiki, the correct command is:

jupyter notebook dig-class100-s2.ipynb

Which is running right now! Exciting stuff!

*EDIT* Ok my first time using all this, so it opens a webpage and you need to:

"click the play button until the end (or click the "double right arrow" to play it all at once)." like @hilsonp noted above

So it seems that with the help of you both I got it running now, fingers crossed :) It's doing now 400 epoch passes and each pass is over a minute so this will take a while :) Seems to be around 8 hours in total on my pc

*EDIT* actually it only runs 80 out of 400 passes so it's already done in less than 2 hours, going to test it now ...

**Comment by gekkehenkie11** [comment]
And .... it doesn't work :( The meter boots but then hangs at "initializing (delayed)..." it says check logs to get more info but there's not much additional info at all.. Any idea @haverland ? here's my file:

https://github.com/gekkehenkie11/android_device_samsung_trltexx/blob/cm-12.0/dig-class100-s2_20250913-161911_q_neu.tflite

I checked and uploaded one of the other files in the output directory and that worked without problems....

Damn, huge bummer after all of this work, hopefully there's some way to get it running

*EDIT* hooking it up to my pc it shows this:

I (29528) TFLITE: Trying to load the model. If it crashes here, it ist most likely due to a corrupted model!
Failed to get registration from op code SHAPE
E (29538) TFLITE: AllocateTensors() failed
Guru Meditation Error: Core  0 panic'ed (LoadProhibited). Exception was unhandled.

Core  0 register dump:
PC      : 0x401b0748  PS      : 0x00060030  A0      : 0x8011c5be  A1      : 0x3ffe8d50  
A2      : 0x3ffbdcac  A3      : 0x00000000  A4      : 0x3ffe8d58  A5      : 0x3ffe8d40  
A6      : 0x3ffe8ca0  A7      : 0x3ffe8d60  A8      : 0x00000000  A9      : 0x00000548  
A10     : 0x0000000e  A11     : 0x00000008  A12     : 0x00000001  A13     : 0x00000013  
A14     : 0x3ffe8be4  A15     : 0x3ffb6f3c  SAR     : 0x00000019  EXCCAUSE: 0x0000001c  
EXCVADDR: 0x00000000  LBEG    : 0x4008a9c6  LEND    : 0x4008a9d1  LCOUNT  : 0x00000000  


Backtrace: 0x401b0745:0x3ffe8d50 0x4011c5bb:0x3ffe8d70 0x4011c87b:0x3ffe8d90 0x40109654:0x3ffe8df0 0x4010978b:0x3ffe8e90 0x40109cdb:0x3ffe8eb0 0x400fbac9:0x3ffe9010 0x40104616:0x3ffe9460 0x40106ba4:0x3ffe94a0




ELF file SHA256: 2a3f258f4

Rebooting...

**Comment by gekkehenkie11** [comment]
Hmm seems some info here: https://www.reddit.com/r/learnmachinelearning/comments/183omdb/didnt_find_op_for_builtin_opcode_shape_version_1/

So I went to https://netron.app/ and opened the model, I see the shape layer but the rest is a bit beyond me I'm afraid, LOL. But as I understand it has to do with incompatible model versions: "The most common issue is that you are using a newer version locally than the hardware supports"

Hopefully all this makes sense to you @haverland and you can guide me to a solution :)

BTW I checked, I have tensorflow 2.20.0 on my pc. Not sure if it's just a matter of downgrading tensorflow and then reconvert that keras model again using a lower tensorflow version?

**Reply by ghost** [reply]
> BTW I checked, I have tensorflow 2.20.0 on my pc. Not sure if it's just a matter of downgrading tensorflow and then reconvert that keras model again using a lower tensorflow version?

tensorflow I have not tested yet. In requirements.txt is tensorflow<=2.17.0. So it should install 2.17.0. This could be the issue

**Reply by ghost** [reply]
py -3.11 -m pip install -r requirements.txt

should install the 2.17.0 version.

**Reply by ghost** [reply]
You can see the used version in notebook under first section "The Model"
```
Tensorflow-Version:  2.17.0
Python: 3.11.11 | packaged by conda-forge | (main, Dec  5 2024, 14:21:42) [Clang 18.1.8 ]
Num GPUs Available:  1
```

**Comment by gekkehenkie11** [comment]
Found this:
https://github.com/haverland/Tenth-of-step-of-a-meter-digit/issues/21

So I did what that guys suggested, now running the notebook again, let's see if that works. Tomorrow :)

**Reply by ghost** [reply]
> Found this: [haverland/Tenth-of-step-of-a-meter-digit#21](https://github.com/haverland/Tenth-of-step-of-a-meter-digit/issues/21)
> 
> So I did what that guys suggested, now running the notebook again, let's see if that works. Tomorrow :)

In https://github.com/haverland/Tenth-of-step-of-a-meter-digit/blob/master/dig-class100-s2.ipynb it is already included unter 

https://github.com/haverland/Tenth-of-step-of-a-meter-digit/blob/69bb445fbd64c91f1776bb9f43b7a1ca87b64dbe/b2n/models/prune_quantize.py#L109

and should work. I'm using visual studio code, but should run with Jupiter notebook alone too.

**Reply by Slider0007** [reply]
Without testing it, I assume keras package needs to be downgraded/limited to 3.10 (keras<=3.10). I was facing the same issue while retraining the analog models recently. It seems for some reason keras 3.11.x (and newer?) includes the SHAPE operator which is not supported by ESP (TFlite micro).

@haverland : I assume all model build environments are affected. I've already changed it here for the analog models ana-clas100 and ana-cont in jomjol repo: https://github.com/jomjol/neural-network-analog-needle-readout/pull/38

I'm not aware if there would be a more elegant solution.

**Reply by caco3** [reply]
> I'm not aware if there would be a more elegant solution.

AFAIK, the way you did it with the requirements file is the way to go

**Reply by SybexX** [reply]
>It seems for some reason keras 3.11.x (and newer?) includes the SHAPE operator which is not supported by ESP (TFlite micro).

@Slider0007 I found something about it, maybe it is helpful.

https://github.com/espressif/esp-tflite-micro/issues/99

https://github.com/tensorflow/tflite-micro/blob/main/tensorflow/lite/micro/micro_mutable_op_resolver.h#L582
https://github.com/jomjol/AI-on-the-edge-device/blob/main/code/components/jomjol_tfliteclass/CTfLiteClass.cpp#L16
https://github.com/jomjol/AI-on-the-edge-device/blob/main/code/components/jomjol_tfliteclass/CTfLiteClass.h#L19

**Reply by Slider0007** [reply]
@SybexX : Thanks for sharing the information. I'm aware that this operator could be added to the ESP, but I was rather interested in a solution of the cause then a solution of the effect. For me it is not clear why the behaviour is so different after only updating a bugfix release of keras. I was already playing with tensor dimensions a bit, e.g. trying to force them static (because model is called as a python function and dimension is a parameter) but was not successful. Therefore I decided to roll back to 3.10 without any obvious drawback (up to now), as the real cause is not clear to me.

Adding the operator to the ESP would be IMHO a breaking change of AIOTE or at least an additional dependency without any? benefit and some downsides. On top the unsupported case is not even transparent to the user, because resulting in an exception without proper log. This would raise a lot of new bug reports over a long period (as long as the user is using older firmware versions with the newest model) with crash reports and you have to answer the same question over and over again ...

**Comment by gekkehenkie11** [comment]
Hey @haverland and @Slider0007 thanks for your help! So i'm now on TF 2.17 and keras 3.10 (I added that line keras<=3.10.0 to the requirements.txt, executed it and then verified with: pip show keras). I cleaned up the build and restarted it again, going to see if it works now, in about 2 hours we'll know :) Thanks!

**Comment by gekkehenkie11** [comment]
Ok it's finished, I loaded it up in https://netron.app/ and it looks way better now, no weird branch anymore. Then I loaded it up in the watermeter and it works indeed, YIHAAAAA !!

Thanks guys!!

So long story short for the next guy trying this: make sure you download the requirements.txt from [@haverland ](https://github.com/haverland/Tenth-of-step-of-a-meter-digit) then add that line:

keras<=3.10.0

and install it

Afterwards you can verify that you indeed have the right versions with:

pip show keras
pip show tensorflow

**Reply by ghost** [reply]
Thanks, will add it to the requirements.txt

**Comment by gekkehenkie11** [comment]
Update: in the last 3 days not even 1 single error!! Really happy with this!!
